### DISCUSSION TITLE
Mailer component and html() call -incompatible with simple_html_dom library ?

### DISCUSSION THREAD

**Discussion by simogeo** [symfony/symfony#54733] [Q&A] [answered]
### Symfony version(s) affected

5.4.38

### Description

I'm trying to switch from _swiftMailer_ to _symfony/mailer_ used as external library ( I don't use symfony)

Everything works well except when I send html manipulated with  `simple_html_dom` library, I get the following error : 

> Fatal error: Uncaught TypeError: The body must be a string, a resource or null (got "simple_html_dom"). in lib/comp-vendor/symfony/mime/Email.php:321 Stack trace: #0 mail/send.php(127): Symfony\Component\Mime\Email->html() #1 plugins/l21_emailSuite/index.php(46): send_email() #2 plugins/index.php(56): include_once('...') #3 templates/admin/2021/index.php(72): include_once('...') #4 admin/index.php(74): include_once('...') #5 {main} thrown in lib/comp-vendor/symfony/mime/Email.php on line 321

When I pass the 

If I send row html (with no `simple_html_dom` library anterior manipulation, I do not experience any issue ... I guesse this is related to mime type ... but I'm stuck. Do you have any clue ? 

Thanks in advance,

### How to reproduce

- prepare a mail with html content using symfony/mailer
- do manipulate html content with `simple_html_dom` library available at http://sourceforge.net/projects/simplehtmldom/
- try to send the mail, it will trigger an error similar to the one provided above

### Additional Context

I know `simple_html_dom` is an old library ! but quite useful ... should symfony/mailer be more permissive ?

**Comment by simogeo** [comment]
By the way, it seems I get a similar result when using https://github.com/voku/simple_html_dom

**Comment by simogeo** [comment, marked as answer]
For your information, a workaround is to cast variable to string 

This generates an error : 
```php
$message = (new Email())
				->from(new Address($from, $fromname))
				->subject($email['subject'])
				->text($email['txt_body'])
				->html($email['html_body']);
```

and this works : 

```php
$message = (new Email())
				->from(new Address($from, $fromname))
				->subject($email['subject'])
				->text($email['txt_body'])
				->html((string) $email['html_body']);
```

**Comment by simogeo** [comment]
The question is : should symfony/mailer try to cast the passed variable itself ?

**Reply by mbabker** [reply]
No, it shouldn't.

The framework tells you the types of information that the method supports and raises an error if you pass in something it doesn't support.  The framework magically casting unsupported types may or may not result in unexpected behavior, so it is better for it to raise an error on an unsupported type.

In your case, the `(string) $email['html_body']` bit works because you're working with an object that has a magic `__toString()` implementation, and what that magic method provides is exactly what you need for your HTML email body.  There is no guarantee the same can be said for other objects with this magic method.

You know what type of information you are working with within the context of your application, and you know what type of information the third-party code you're calling needs.  Your application code is best able to handle converting data into what the third-party code needs versus relying on the third-party code to try and create something that may or may not be appropriate for its use.
